File: wmdemo/wmdemo.c
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>

#include <cwiid.h>

/* This is a sample program written to demonstrate basic CWiid libwiimote
 * usage, until _actual_ documentation can be written.  It's quick and dirty
 * has a horrible interface, but it's sparce enough to pick out the important
 * parts easily.  For examples of read and write code, see wmgui.  Speaker
 * support is "experimental" (read: bad) enough to be disabled.  The beginnings
 * of a speaker output function are in libwiimote source. */
/* Note: accelerometer (including nunchuk) and IR outputs produce a
 * lot of data - the purpose of this program is demonstration, not good
 * interface, and it shows. */

cwiid_mesg_callback_t cwiid_callback;

#define toggle_bit(bf,b)	\
	(bf) = ((bf) & b)		\
	       ? ((bf) & ~(b))	\
	       : ((bf) | (b))

void set_led_state(cwiid_wiimote_t *wiimote, unsigned char led_state);
void set_rpt_mode(cwiid_wiimote_t *wiimote, unsigned char rpt_mode);

cwiid_err_t err;
void err(int id, const char *s, ...)
{
	va_list ap;

	va_start(ap, s);
	printf("%d:", id);
	vprintf(s, ap);
	printf("\n");
	va_end(ap);
}

/* wiimote handle */
cwiid_wiimote_t *wiimote;

int main(int argc, char *argv[])
{
	bdaddr_t bdaddr;	/* bluetooth device address */
	int cwiid_id;		/* wiimote id: useful for handling multiple wiimotes
	                       with a single callback */
	unsigned char led_state = 0;
	unsigned char rpt_mode = 0;
	unsigned char rumble = 0;
	int exit = 0;
	char *str_addr;
	int c;

	cwiid_set_err(err);

	
	/* Connect to address given on command-line, if present */
	if (argc > 1) {
		str2ba(argv[1], &bdaddr);
	}
	else {
		bdaddr = *BDADDR_ANY;
	}

	/* Connect to the wiimote */
	printf("Put Wiimote in discoverable mode now (press 1+2)...\n");
	if (!(wiimote = cwiid_connect(&bdaddr, cwiid_callback, &cwiid_id))) {
		fprintf(stderr, "Unable to connect to wiimote\n");
		return -1;
	}

	/* Menu */
	printf("1: toggle LED 1\n"
	       "2: toggle LED 2\n"
	       "3: toggle LED 3\n"
	       "4: toggle LED 4\n"
	       "a: toggle accelerometer output\n"
	       "b: toggle button output\n"
	       "e: toggle extension output\n"
	       "i: toggle ir output\n"
	       "r: toggle rumble\n"
	       "s: request status message (use t to turn on output)\n"
	       "t: toggle status output\n"
	       "x: exit\n");

	while (!exit) {
		switch (getchar()) {
		case '1':
			toggle_bit(led_state, CWIID_LED1_ON);
			set_led_state(wiimote, led_state);
			break;
		case '2':
			toggle_bit(led_state, CWIID_LED2_ON);
			set_led_state(wiimote, led_state);
			break;
		case '3':
			toggle_bit(led_state, CWIID_LED3_ON);
			set_led_state(wiimote, led_state);
			break;
		case '4':
			toggle_bit(led_state, CWIID_LED4_ON);
			set_led_state(wiimote, led_state);
			break;
		case 'a':
			toggle_bit(rpt_mode, CWIID_RPT_ACC);
			set_rpt_mode(wiimote, rpt_mode);
			break;
		case 'b':
			toggle_bit(rpt_mode, CWIID_RPT_BTN);
			set_rpt_mode(wiimote, rpt_mode);
			break;
		case 'e':
			/* CWIID_RPT_EXT is actually
			 * CWIID_RPT_NUNCHUK | CWIID_RPT_CLASSIC */
			toggle_bit(rpt_mode, CWIID_RPT_EXT);
			set_rpt_mode(wiimote, rpt_mode);
			break;
		case 'i':
			/* libwiimote picks the highest quality IR mode available with the
			 * other options selected (not including as-yet-undeciphered
			 * interleaved mode */
			toggle_bit(rpt_mode, CWIID_RPT_IR);
			set_rpt_mode(wiimote, rpt_mode);
			break;
		case 'r':
			toggle_bit(rumble, 1);
			if (cwiid_command(wiimote, CWIID_CMD_RUMBLE, rumble)) {
				fprintf(stderr, "Error setting rumble\n");
			}
			break;
		case 's':
			if (cwiid_command(wiimote, CWIID_CMD_STATUS, 0)) {
				fprintf(stderr, "Error requesting status message\n");
			}
			break;
		case 't':
			toggle_bit(rpt_mode, CWIID_RPT_STATUS);
			set_rpt_mode(wiimote, rpt_mode);
			break;
		case 'x':
			exit = -1;
			break;
		case '\n':
			break;
		default:
			fprintf(stderr, "invalid option\n");
		}
	}

	if (cwiid_disconnect(wiimote)) {
		fprintf(stderr, "Error on wiimote disconnect\n");
		return -1;
	}

	return 0;
}

void set_led_state(cwiid_wiimote_t *wiimote, unsigned char led_state)
{
	if (cwiid_command(wiimote, CWIID_CMD_LED, led_state)) {
		fprintf(stderr, "Error setting LEDs \n");
	}
}
	
void set_rpt_mode(cwiid_wiimote_t *wiimote, unsigned char rpt_mode)
{
	if (cwiid_command(wiimote, CWIID_CMD_RPT_MODE, rpt_mode)) {
		fprintf(stderr, "Error setting report mode\n");
	}
}

/* Prototype cwiid_callback with cwiid_callback_t, define it with the actual
 * type - this will cause a compile error (rather than some undefined bizarre
 * behavior) if cwiid_callback_t changes */
/* cwiid_mesg_callback_t has undergone a few changes lately, hopefully this
 * will be the last.  Some programs need to know which messages were received
 * simultaneously (e.g. for correlating accelerometer and IR data), and the
 * sequence number mechanism used previously proved cumbersome, so we just
 * pass an array of messages, all of which were received at the same time.
 * The id is to distinguish between multiple wiimotes using the same callback.
 * */
void cwiid_callback(int id, int mesg_count, union cwiid_mesg *mesg[])
{
	int i, j;
	int valid_source;

	for (i=0; i < mesg_count; i++)
	{
		switch (mesg[i]->type) {
		case CWIID_MESG_STATUS:
			printf("Status Report: battery=%d extension=",
			       mesg[i]->status_mesg.battery);
			switch (mesg[i]->status_mesg.extension) {
			case CWIID_EXT_NONE:
				printf("none");
				break;
			case CWIID_EXT_NUNCHUK:
				printf("Nunchuk");
				break;
			case CWIID_EXT_CLASSIC:
				printf("Classic Controller");
				break;
			default:
				printf("Unknown Extension");
				break;
			}
			printf("\n");
			break;
		case CWIID_MESG_BTN:
			printf("Button Report: %.4X\n", mesg[i]->btn_mesg.buttons);
			break;
		case CWIID_MESG_ACC:
			printf("Acc Report: x=%d, y=%d, z=%d\n", mesg[i]->acc_mesg.x,
			                                         mesg[i]->acc_mesg.y,
			                                         mesg[i]->acc_mesg.z);
			break;
		case CWIID_MESG_IR:
			printf("IR Report: ");
			valid_source = 0;
			for (j = 0; j < CWIID_IR_SRC_COUNT; j++) {
				if (mesg[i]->ir_mesg.src[j].valid) {
					valid_source = 1;
					printf("(%d,%d) ", mesg[i]->ir_mesg.src[j].x,
					                   mesg[i]->ir_mesg.src[j].y);
				}
			}
			if (!valid_source) {
				printf("no sources detected");
			}
			printf("\n");
			break;
		case CWIID_MESG_NUNCHUK:
			printf("Nunchuk Report: btns=%.2X stick=(%d,%d) acc.x=%d acc.y=%d "
			       "acc.z=%d\n", mesg[i]->nunchuk_mesg.buttons,
			       mesg[i]->nunchuk_mesg.stick_x,
			       mesg[i]->nunchuk_mesg.stick_y, mesg[i]->nunchuk_mesg.acc_x,
			       mesg[i]->nunchuk_mesg.acc_y, mesg[i]->nunchuk_mesg.acc_z);
			break;
		case CWIID_MESG_CLASSIC:
			printf("Classic Report: btns=%.4X l_stick=(%d,%d) r_stick=(%d,%d) "
			       "l=%d r=%d\n", mesg[i]->classic_mesg.buttons,
			       mesg[i]->classic_mesg.l_stick_x,
			       mesg[i]->classic_mesg.l_stick_y,
			       mesg[i]->classic_mesg.r_stick_x,
			       mesg[i]->classic_mesg.r_stick_y,
			       mesg[i]->classic_mesg.l, mesg[i]->classic_mesg.r);
			break;
		case CWIID_MESG_ERROR:
			if (cwiid_disconnect(wiimote)) {
				fprintf(stderr, "Error on wiimote disconnect\n");
				exit(-1);
			}
			exit(0);
			break;
		default:
			printf("Unknown Report");
			break;
		}
	}
}

File: wmdemo/wmdemo.py
#!/usr/bin/python
import cwiid
import sys

menu = '''1: toggle LED 1
2: toggle LED 2
3: toggle LED 3
4: toggle LED 4
5: toggle rumble
a: toggle accelerometer reporting
b: toggle button reporting
e: toggle extension reporting
i: toggle ir reporting
m: toggle messages
p: print this menu
r: request status message ((t) enables callback output)
s: print current state
t: toggle status reporting
x: exit'''

def main():
	led = 0
	rpt_mode = 0
	rumble = 0
	mesg = False

	#Connect to address given on command-line, if present
	print 'Put Wiimote in discoverable mode now (press 1+2)...'
	if len(sys.argv) > 1:
		wiimote = cwiid.Wiimote(sys.argv[1])
	else:
		wiimote = cwiid.Wiimote()

	wiimote.mesg_callback = callback

	print menu

	exit = 0
	while not exit:
		c = sys.stdin.read(1)
		if c == '1':
			led ^= cwiid.LED1_ON
			wiimote.led = led
		elif c == '2':
			led ^= cwiid.LED2_ON
			wiimote.led = led
		elif c == '3':
			led ^= cwiid.LED3_ON
			wiimote.led = led
		elif c == '4':
			led ^= cwiid.LED4_ON
			wiimote.led = led
		elif c == '5':
			rumble ^= 1
			wiimote.rumble = rumble
		elif c == 'a':
			rpt_mode ^= cwiid.RPT_ACC
			wiimote.rpt_mode = rpt_mode
		elif c == 'b':
			rpt_mode ^= cwiid.RPT_BTN
			wiimote.rpt_mode = rpt_mode
		elif c == 'e':
			rpt_mode ^= cwiid.RPT_EXT
			wiimote.rpt_mode = rpt_mode
		elif c == 'i':
			rpt_mode ^= cwiid.RPT_IR
			wiimote.rpt_mode = rpt_mode
		elif c == 'm':
			mesg = not mesg
			if mesg:
				wiimote.enable(cwiid.FLAG_MESG_IFC);
			else:
				wiimote.disable(cwiid.FLAG_MESG_IFC);
		elif c == 'p':
			print menu
		elif c == 'r':
			wiimote.request_status()
		elif c == 's':
			print_state(wiimote.state)
		elif c == 't':
			rpt_mode ^= cwiid.RPT_STATUS
			wiimote.rpt_mode = rpt_mode
		elif c == 'x':
			exit = -1;
		elif c == '\n':
			pass
		else:
			print 'invalid option'

	wiimote.close()

def print_state(state):
	print 'Report Mode:',
	for r in ['STATUS', 'BTN', 'ACC', 'IR', 'NUNCHUK', 'CLASSIC']:
		if state['rpt_mode'] & eval('cwiid.RPT_' + r):
			print r,
	print

	print 'Active LEDs:',
	for led in ['1','2','3','4']:
		if state['led'] & eval('cwiid.LED' + led + '_ON'):
			print led,
	print

	print 'Rumble:', state['rumble'] and 'On' or 'Off'

	print 'Battery:', int(100.0 * state['battery'] / cwiid.BATTERY_MAX)

	if 'buttons' in state:
		print 'Buttons:', state['buttons']

	if 'acc' in state:
		print 'Acc: x=%d y=%d z=%d' % (state['acc'][cwiid.X],
		                               state['acc'][cwiid.Y],
		                               state['acc'][cwiid.Z])

	if 'ir_src' in state:
		valid_src = False
		print 'IR:',
		for src in state['ir_src']:
			if src:
				valid_src = True
				print src['pos'],

		if not valid_src:
			print 'no sources detected'
		else:
			print

	if state['ext_type'] == cwiid.EXT_NONE:
		print 'No extension'
	elif state['ext_type'] == cwiid.EXT_UNKNOWN:
		print 'Unknown extension attached'
	elif state['ext_type'] == cwiid.EXT_NUNCHUK:
		print 'Nunchuk: btns=%.2X stick=%r acc.x=%d acc.y=%d acc.z=%d' % \
		  (state['nunchuk']['buttons'], state['nunchuk']['stick'],
		   state['nunchuk']['acc'][cwiid.X],
		   state['nunchuk']['acc'][cwiid.Y],
		   state['nunchuk']['acc'][cwiid.Z])
	elif state['ext_type'] == cwiid.EXT_CLASSIC:
		print 'Classic: btns=%.4X l_stick=%r r_stick=%r l=%d r=%d' % \
		  (state['classic']['buttons'],
		   state['classic']['l_stick'], state['classic']['r_stick'],
		   state['classic']['l'], state['classic']['r'])

def callback(mesg_list):
	for mesg in mesg_list:
		if mesg[0] == cwiid.MESG_STATUS:
			print 'Status Report: battery=%d extension=' % \
			       mesg[1]['battery'],
			if mesg[1]['ext_type'] == cwiid.EXT_NONE:
				print 'none'
			elif mesg[1]['ext_type'] == cwiid.EXT_NUNCHUK:
				print 'Nunchuk'
			elif mesg[1]['ext_type'] == cwiid.EXT_CLASSIC:
				print 'Classic Controller'
			else:
				print 'Unknown Extension'

		elif mesg[0] == cwiid.MESG_BTN:
			print 'Button Report: %.4X' % mesg[1]

		elif mesg[0] == cwiid.MESG_ACC:
			print 'Acc Report: x=%d, y=%d, z=%d' % \
			      (mesg[1][cwiid.X], mesg[1][cwiid.Y], mesg[1][cwiid.Z])

		elif mesg[0] == cwiid.MESG_IR:
			valid_src = False
			print 'IR Report: ',
			for src in state['ir_src']:
				if src:
					valid_src = True
					print src['pos'],

			if not valid_src:
				print 'no sources detected'
			else:
				print

		elif mesg[0] == cwiid.MESG_NUNCHUK:
			print 'Nunchuk Report: btns=%.2X stick=%r ' + \
			      'acc.x=%d acc.y=%d acc.z=%d' % (state['nunchuk']['buttons'],
			   state['nunchuk']['stick'],
			   state['nunchuk']['acc'][cwiid.X],
			   state['nunchuk']['acc'][cwiid.Y],
			   state['nunchuk']['acc'][cwiid.Z])
		elif mesg[0] == cwiid.MESG_CLASSIC:
			print 'Classic Report: btns=%.4X l_stick=%r r_stick=%r' + \
			      'l=%d r=%d' % (state['classic']['buttons'],
			   state['classic']['l_stick'], state['classic']['r_stick'],
			   state['classic']['l'], state['classic']['r'])
		elif mesg[0] ==  cwiid.MESG_ERROR:
			wiimote.close()
			exit(-1)
		else:
			print 'Unknown Report'

main()
